libraries/AP_GPS/AP_GPS.cpp

```diff
@@ -527,47 +557,70 @@ AP_GPS::update(void)
         update_instance(i);
     }
 
-    // work out which GPS is the primary, and how many sensors we have
-    for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
-        if (state[i].status != NO_GPS) {
-            num_instances = i+1;
-        }
-        if (_auto_switch) {            
-            if (i == primary_instance) {
-                continue;
-            }
-            if (state[i].status > state[primary_instance].status) {
-                // we have a higher status lock, change GPS
-                primary_instance = i;
-                continue;
+    // Implement soft switch logic
+    if (_auto_switch == 2) {
+        // calculate weighting for each GPS
+        calc_blend_weights();
+
+        // Use the weighting to calculate blended GPS states
+        calc_blended_state();
+
+        // update notify with gps status. Used blended data
+        AP_Notify::flags.gps_status = state[GPS_MAX_INSTANCES].status;
+        AP_Notify::flags.gps_num_sats = state[GPS_MAX_INSTANCES].num_sats;
+
+    } else {
+        // use hard switch logic
+        // work out which GPS is the primary, and how many sensors we have
+        for (uint8_t i=0; i<GPS_MAX_INSTANCES; i++) {
+            if (state[i].status != NO_GPS) {
+                num_instances = i+1;
             }
+            if (_auto_switch) {
+                if (i == primary_instance) {
+                    continue;
+                }
+                if (state[i].status > state[primary_instance].status) {
+                    // we have a higher status lock, change GPS
+                    primary_instance = i;
```

Comment:
This is not really a new fault, but the last_instance_swap_ms should be updated here.